### PR TITLE
[WIP] Text templates

### PR DESCRIPTION
**Pull request by gratimax**
See https://gist.github.com/gratimax/9aa7ecc33467e0ddfd0b. More info coming soon.

Hopefully this example can showcase some of the intended power here:

```java
import static org.spongepowered.api.text.template.TextElements.*;

static final TextTemplate playerMessage = TextTemplate.of(
  TextColors.BLUE, "Hello, ", TextColors.LIGHT_GRAY,
  playerDisplayName().pos(0), TextColors.BLUE,
  "See your team members: ", 
  TextColors.LIGHT_GRAY, 
    fallback(
      iterable(playerDisplayName(), Texts.of("|")), 
      Texts.of("No players")).pos(1), 
  TextColors.BLUE, "!");

for (Player player: players) {
  player.sendMessage(playerMessage.build(
    TextArgs.of(player, players)));
}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/api/text/Text.java [hunk 1]
```diff
@@ -289,6 +285,12 @@ public final String getContent() {
         }
 
         @Override
+        public boolean isEmpty() {
+            return content.equals("") && children.isEmpty() && !clickAction.isPresent() && !hoverAction.isPresent() &&
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/text/Text.java]: You could reduce this to 
````java
return super.isEmpty() && this.content.isEmpty();
````

**Comment by TheRaspPie** [comment]
Will this remove `Texts.format`? Because if yes, we couldn't do a per-player message with `MessageSink.transformMessage` anymore.
If no, then we have two formatting APIs.

**Comment by gratimax** [comment]
@TheRaspPie 

Yes, it will remove `Texts.format`. I don't really get what you mean about per-player messaging, here it is in action:
```java
class MyMessageSink extends MessageSink {
   private final TextTemplate template = TextTemplate.of("A message was just sent to you: ", TextElement.identity().pos(0));

   public Text transformMessage(CommandSource target, Text text) {
        return template.build(TextArgs.of(text));
    }
}
```

I don't really get how having to pass around text templates is "inflexible". If anything, it makes using the text API clean since there is a really clean separation of concerns.

**Comment by Flibio** [comment]
@gratimax 

Any idea when this PR will be accepted?

**Comment by gratimax** [comment]
@Flibio Sorry, the PR is still not done (there's quite a lot to do yet), so I cannot guarantee any ETA. Sorry for making you wait :frowning:. Is there a specific use case you have that could perhaps be done sooner, so that you won't wait on this PR?

**Comment by Zidane** [comment]
@gratimax 

MessageEvent needs template support to do formatting without touching source names. Really need to get this sped along into master.

**Comment by gratimax** [comment]
@Zidane I understand that. I suggested in the comment that the specific use case of MessageEvent's touching source names could be done sooner so that no one has to wait on this PR. My suggestion would be to use functions representing templates, which is ugly but works provisionally. I appreciate that people have a use for this, but I can't guarantee any time that this PR can be merged since there is still a lot to do, as I have said several times before.

**Comment by ST-DDT** [comment]
You could use Texts.format for that as well.

**Comment by Zidane** [comment]
@gratimax 

I'd rather it be done right and, barring additional work. this PR is pretty nice. All I'm saying is it needs to become a priority.

**Comment by Flibio** [comment]
@gratimax 

I'm just waiting for the ability to get the raw message from the `MessageSinkEvent`. I need the plain message, without the player name included.

**Comment by bloodmc** [comment]
@gratimax Any idea when this will be finished?

**Comment by Wolfizen** [comment]
I think the best feature of this PR is separating placeholders from text.

**Comment by ryantheleach** [comment]
@Wolfizen +1 It also solves the issue of having multiple matching "variables" in shoddy message templating (but that could have been solved by using a message templating language that has starting and ending tags / brackets

e.g. +g and +global matching (g)lobal making Administratorlobal

File: src/main/java/org/spongepowered/api/text/template/TextArgs.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    private TextArgs(ImmutableList.Builder<Object> positionedArgsBuilder,
+                     ImmutableMap.Builder<String, Object> namedArgsBuilder) {
+        this.positionedArgsBuilder = positionedArgsBuilder;
+        this.namedArgsBuilder = namedArgsBuilder;
+    }
+
+    public TextArgs add(Object arg) {
+        if (positionedArgsBuilder == null) {
+            throw new IllegalArgumentException("Positioned arguments were already accessed, to add" +
+                "more arguments use TextArgs.copy()");
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/text/template/TextArgs.java]: Should have a space after add

File: src/main/java/org/spongepowered/api/text/template/TextArgs.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        positionedArgsBuilder.add(arg);
+        return this;
+    }
+
+    public TextArgs add(Object... args) {
+        if (positionedArgsBuilder == null) {
+            throw new IllegalArgumentException("Positioned arguments were already accessed, to add" +
+                "more arguments use TextArgs.copy()");
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/text/template/TextArgs.java]: Another space after add

File: src/main/java/org/spongepowered/api/text/template/TextArgs.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        positionedArgsBuilder.add(args);
+        return this;
+    }
+
+    public TextArgs with(String name, Object arg) {
+        if (positionedArgsBuilder == null) {
+            throw new IllegalArgumentException("Named arguments were already accessed, to add" +
+                "more arguments use TextArgs.copy()");
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/text/template/TextArgs.java]: Another space after add

File: src/main/java/org/spongepowered/api/text/template/TextElements.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                return text;
+            }
+        };
+    }
+
+    public static <T> TextElement<T> function(final Function<T, Text> function) {
+        return new TextElement<T>() {
+            @Override
+            public Text create(T value) {
+                return Preconditions.checkNotNull(function.apply(value));
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/text/template/TextElements.java]: I think `Preconditions` is meant to be used via static imports, right?

Comment (simon816) [inline src/main/java/org/spongepowered/api/text/template/TextElements.java]: Yes

**Comment by ST-DDT** [comment]
Sorry for repeating myself, but are there any thoughts about recreating the TextRepresentable feature? (Show Uuid for player and start messaging him on click, show the ItemStack/Achievement details in chat. With all plugins doing it the same way on a server without any knowledge how that impl might display them. And especially not limited to API provided ones.)

Have there been considerations on how to handle the (de-)serialization of the TextTemplates?

**Comment by gratimax** [comment]
The TextRepresentable feature here is no longer useful, since you can simply use the relevant TextElement. This has many advantages, such as allowing for multiple representations of any object and one clear place to see all the representations. I have added a few elements that mimic TextRepresentable, but the other ones haven't been added yet.

I have thought about deserialization. It is certainly possible, but it is out of scope, and I think way too opinionated to fit into an API.

**Comment by ST-DDT** [comment]
> I have thought about deserialization. It is certainly possible, but it is out of scope, and way too opinionated to fit into an API. With the TextElement/TextTemplate framework, however, it is easy for a plugin to provide its own deserialization.

IMO its a must have. Its probably one of the first things I will write for my plugins. And I would be glad if the API would do that for me (even if I have to write it for the API). The entire use of being able to load messages from json/file is directly connected to the ability to replace variables in messages loaded from there.

> Previously with placeholders, this was not possible, since they were mixed in with normal text objects.

Thats true, but has been fixed in the latest PR.

**Comment by gratimax** [comment]
Just because text configuration is something that _you_ are going to write for your plugins doesn't mean that it should be a utility that is forced upon everyone using the API. "Serializing" templates goes straight to dealing with configuration, which is 100% out of scope with both this PR and the API in general. Let plugin developers decide how configuration will be stored.

**Comment by boformer** [comment]
I disagree. The deserialization of templates should be standardized. That's very important for i18n.

**Comment by simon816** [comment]
Could this be merged into #808 or at least a decision reached between the two because nothing seems to be happening at the moment

**Comment by ST-DDT** [comment]
@simon816
Which features should be ported/merged over to my PR? AFAIK its possible to do everything proposed by this PR with mine already.
I will have an in detail look on the features of this PR this weekend to make sure I did not miss anything.

PS: My PR is almost up to date with master. I just have to update my testplugin and check some javadocs this weekend.

**Comment by simon816** [comment]
The fact that this PR is still open suggests that this is PR is still to be considered.
From what @Minecrell [said on #808](https://github.com/SpongePowered/SpongeAPI/pull/808#issuecomment-153811391)
> It's waiting on completion of #835 so we can decide which one to use.

I think @Minecrell / @gratimax should check over similarities/differences as I don't have an opinion on this topic.

File: src/main/java/org/spongepowered/api/text/template/TextArgs.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+    /**
+     * Adds the given arguments to end of the positioned arguments.
+     *
+     * @param args The new arguments
+     * @return A new TextArgs object
+     */
+    public TextArgs add(Object... args) {
+        return new TextArgs(
+            ImmutableList.builder().addAll(positionedArgs).add(args).build(),
```
Comment (TheRaspPie) [inline src/main/java/org/spongepowered/api/text/template/TextArgs.java]: Shouldn't this be `addAll(positionedArgs).addAll(args)`?

**Comment by gratimax** [comment]
Closing this because I don't have enough time to finish. Sorry about that!

**Pull request closed without merging**